### PR TITLE
Enable HOMEBREW_PATCHELF_RB_WRITE for CI and HOMEBREW_DEV_CMD_RUN

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----
    
Enable `HOMEBREW_PATCHELF_RB_WRITE` by default for `HOMEBREW_DEV_CMD_RUN` and `CI`, but not when building bottles.

It may be disabled using `HOMEBREW_NO_PATCHELF_RB_WRITE`.
It may be enabled on CI using `HOMEBREW_PATCHELF_RB_WRITE`.

Fixes https://github.com/Homebrew/linuxbrew-core/issues/21601

### CODE REVIEW

**Comment by sjackman** [comment]
Thanks to @rmNULL who implemented `HOMEBREW_PATCHELF_RB_WRITE`.

**Review by woodruffw — APPROVED** [review]
❤️

**Review by MikeMcQuaid — COMMENTED** [review]
> except CI.

Why except CI? I think if we're going to enable it for normal users we should enable it for CI too. Arguably we should enable for CI before normal users otherwise I don't see us catching issues before this ends up in a stable tag and affects a lot of people.

**Comment by sjackman** [comment]
> Why except CI? I think if we're going to enable it for normal users we should enable it for CI too. Arguably we should enable for CI before normal users otherwise I don't see us catching issues before this ends up in a stable tag and affects a lot of people.

Because CI builds bottles, and if it were to create bottles that were subtly broken but we didn't realize until later, that'd be bad and harder to fix. We could enable `HOMEBREW_PATCHELF_RB_WRITE` by default for CI, and disable it by default in `brew bottle`.

**Comment by MikeMcQuaid** [comment]
> We could enable `HOMEBREW_PATCHELF_RB_WRITE` by default for CI, and disable it by default in `brew bottle`.

This seems like a reasonable compromise and could be done in `brew test-bot`.

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 1]
```diff
@@ -84,6 +84,8 @@ def bottle_args
   end
 
   def bottle
+    ENV["HOMEBREW_NO_PATCHELF_RB_WRITE"] = "1" unless ENV["HOMEBREW_PATCHELF_RB_WRITE"].present?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bottle.rb:87]: Was thinking this could go into `brew test-bot`?

File: Library/Homebrew/os/linux/elf.rb [hunk 2]
```diff
@@ -101,7 +101,7 @@ def interpreter
   def patch!(interpreter: nil, rpath: nil)
     return if interpreter.blank? && rpath.blank?
 
-    if HOMEBREW_PATCHELF_RB_WRITE
+    if HOMEBREW_PATCHELF_RB_WRITE && ENV["HOMEBREW_NO_PATCHELF_RB_WRITE"].blank?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux/elf.rb:104]: Can remove this change if this is set out of process (which I think would be a bit less confusing).

File: Library/Homebrew/os/linux/global.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@
 # Enables experimental `patchelf.rb` write support.
 HOMEBREW_PATCHELF_RB_WRITE = (
   ENV["HOMEBREW_NO_PATCHELF_RB_WRITE"].blank? &&
-  (ENV["HOMEBREW_PATCHELF_RB_WRITE"].present? || ENV["CI"].blank?)
+  (ENV["HOMEBREW_PATCHELF_RB_WRITE"].present? || ENV["CI"].present? || ENV["HOMEBREW_DEV_CMD_RUN"].present?)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux/global.rb:7]: I think this could be limited just to `HOMEBREW_NO_PATCHELF_RB_WRITE.blank?` if this is moved to test-bot.

**Comment by sjackman** [comment]
PR https://github.com/Homebrew/homebrew-test-bot/pull/525 must be merged before this PR.

File: Library/Homebrew/os/linux/global.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/global.rb:7]: I'd prefer to roll it out to CI and HOMEBREW_DEV_CMD_RUN first, and give it one stable release to burn in, before rolling it out to everyone.

File: Library/Homebrew/dev-cmd/bottle.rb (hunk 1 above)
Comment (sjackman) [reply Library/Homebrew/dev-cmd/bottle.rb]: Done!

File: Library/Homebrew/os/linux/elf.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/os/linux/elf.rb]: Done!

**Comment by sjackman** [comment]
PR https://github.com/Homebrew/homebrew-test-bot/pull/525 has been merged.

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**